torch/csrc/api/src/nn/modules/transformer.cpp

```diff
@@ -342,5 +326,110 @@ Tensor TransformerDecoderImpl::forward(
   return output;
 }
 
+
+// =======================================TransformerImpl================================
+TransformerImpl::TransformerImpl(TransformerOptions options_ ) : options(std::move(options_)){
+  reset();
+}
+
+void TransformerImpl::reset() {
+  // set up encoder
+  if (options.custom_encoder().is_empty()) {
+    LayerNorm norm(LayerNormOptions({options.d_model()}));
+    TransformerEncoder trans_encoder(TransformerEncoderOptions(
+      TransformerEncoderLayerOptions(options.d_model(), options.nhead())
+      .dim_feedforward(options.dim_feedforward())
+      .dropout(options.dropout())
+      .activation(options.activation()), options.num_encoder_layers())
+        .norm(AnyModule(norm)));
+
+    this->encoder = AnyModule(trans_encoder);
+  }
+  else {
+    this->encoder = options.custom_encoder().clone();
+  }
+  this->register_module("encoder", this->encoder.ptr());
+
+  // set up decoder
+  if (options.custom_decoder().is_empty()) {
+
+    LayerNorm norm(LayerNormOptions({options.d_model()}));
+    TransformerDecoder trans_decoder(TransformerDecoderOptions(
+      TransformerDecoderLayerOptions(options.d_model(), options.nhead())
+      .dim_feedforward(options.dim_feedforward())
+      .dropout(options.dropout())
+      .activation(options.activation()), options.num_decoder_layers())
+        .norm(AnyModule(norm)));
+
+    this->decoder = AnyModule(trans_decoder);
+  }
+  else {
+    this->decoder = options.custom_decoder().clone();
+  }
+  this->register_module("decoder", this->decoder.ptr());
+
+  reset_parameters();
+}
+
+
+void TransformerImpl::reset_parameters() {
+  auto parameters = this->parameters();
```

Comment:
Should probably be `auto&`, you don't want to (shallow) copy the vector, https://github.com/pytorch/pytorch/blob/367426494759ddde0896665ed55c6f9af2870cf0/torch/csrc/api/include/torch/nn/module.h#L208

Reply:
Actually, nevermind, I see that there is no API to get a reference to the parameters like that